blog/2020-10-01-hs-to-coq/2020-10-01-hs-to-coq.md

```diff
@@ -0,0 +1,1084 @@
+  rather than applied to the types of its fields
+  (as in `MkQueue [a] [a]`).
+
+```coq
+Inductive Queue a : Type := | MkQueue : list a -> list a -> Queue a.
+```
+
+In Coq, identifiers for functions (which includes data constructors)
+can have some of their arguments implicit. The following command
+makes the first argument of `MkQueue` implicit. Note that `MkQueue`
+actually has three arguments: the type `a`, and the two lists.
+The type `a` can be inferred from the other two, hence it is desirable
+to make it implicit.
+
+```coq
+Arguments MkQueue {_} _ _.
+```
+
+The rest are definitions.
+
+- Type parameters such as `a` are actually regular parameters to the function.
+- `match` expressions can match on multiple values simultaneously, which makes
+  it convenient to translate multi-clause definitions.
+- Some generated names `arg_0__` and `arg_1__` stand for the arguments of the
+  function just before we pattern-match on them. Each branch reuses the names
+  from the Haskell source.
+
+```coq
+Definition push {a} : a -> Queue a -> Queue a :=
+  fun arg_0__ arg_1__ =>
+    match arg_0__, arg_1__ with
+    | x, MkQueue xs ys => MkQueue (cons x xs) ys
+    end.
+```
+
+<div class="tuto">
+<details>
+<summary>Complete Coq module `Queue.v`</summary>
+```coq
+(* Default settings (from HsToCoq.Coq.Preamble) *)
+
+Generalizable All Variables.
+
+Unset Implicit Arguments.
+Set Maximal Implicit Insertion.
+Unset Strict Implicit.
+Unset Printing Implicit Defensive.
+
+Require Coq.Program.Tactics.
+Require Coq.Program.Wf.
+
+(* Converted imports: *)
+
+Require GHC.List.
+
+(* Converted type declarations: *)
+
+Inductive Queue a : Type := | MkQueue : list a -> list a -> Queue a.
+
+Arguments MkQueue {_} _ _.
+
+(* Converted value declarations: *)
+
+Definition push {a} : a -> Queue a -> Queue a :=
+  fun arg_0__ arg_1__ =>
+    match arg_0__, arg_1__ with
+    | x, MkQueue front back => MkQueue front (cons x back)
+    end.
+
+Definition pop {a} : Queue a -> option (a * Queue a)%type :=
+  fun arg_0__ =>
+    match arg_0__ with
+    | MkQueue (cons y front) back => Some (pair y (MkQueue front back))
+    | MkQueue nil back =>
+        match GHC.List.reverse back with
+        | nil => None
+        | cons y front => Some (pair y (MkQueue front nil))
+        end
+    end.
+
+Definition empty {a} : Queue a :=
+  MkQueue nil nil.
+
+(* External variables:
+     None Some cons list nil op_zt__ option pair GHC.List.reverse
+*)
+```
+</details>
+</div>
+
+### Verification
+
+Correctness is relative to a particular *specification*:
+the implementation is "correct" when it conforms to the specification.
+In this section, we will define a specification for the queue,
+and prove that the queue implements that specification.
+
+For a data structure such as a queue, a common form of specification is to
+relate it to a simpler *reference implementation*, or *model*, which is
+considered "obviously correct".
```

Comment:
Might worth explaining a little bit more what is "obviously correct". I imagine something you write in the discussion can be foreshadowed here to help to explain that.